### PR TITLE
[ASTextNode] Further locking improvements for ASTextNode

### PR DESCRIPTION
**Pull request by maicki**

### CODE REVIEW

File: AsyncDisplayKit/ASTextNode.mm [hunk 1]
```diff
@@ -394,26 +400,23 @@ - (void)setAttributedText:(NSAttributedString *)attributedText
 
   // Tell the display node superclasses that the cached layout is incorrect now
   [self invalidateCalculatedLayout];
-
```
Comment (appleguy) [inline AsyncDisplayKit/ASTextNode.mm]: I wonder if we need to include the renderer invalidation and the setNeedsDisplay in the lock context?  If we don't, what happens if another thread writes to other properties in the meantime?

File: AsyncDisplayKit/ASTextNode.mm [hunk 2]
```diff
@@ -394,26 +400,23 @@ - (void)setAttributedText:(NSAttributedString *)attributedText
...
-  if (ASObjectIsEqual(exclusionPaths, _exclusionPaths)) {
-    return;
+    if (ASObjectIsEqual(exclusionPaths, _exclusionPaths)) {
+      return;
+    }
+  
+    _exclusionPaths = [exclusionPaths copy];
   }
   
-  _exclusionPaths = [exclusionPaths copy];
```
Comment (appleguy) [inline AsyncDisplayKit/ASTextNode.mm]: Same implication here regarding renderer and display invalidation, these operations may need to occur together as if they were a "transaction".

File: AsyncDisplayKit/ASTextNode.mm [hunk 3]
```diff
@@ -713,6 +713,8 @@ - (void)_setHighlightRange:(NSRange)highlightRange forAttributeName:(NSString *)
       }
 
       if (highlightTargetLayer != nil) {
+        std::lock_guard<std::recursive_mutex> l(_textLock);
```
Comment (appleguy) [inline AsyncDisplayKit/ASTextNode.mm]: So the lock here was missing?

Comment (maicki) [inline AsyncDisplayKit/ASTextNode.mm]: Yes for the renderer

**Comment by appleguy** [comment]
@maicki This looks great to me!

**Pull request merged**